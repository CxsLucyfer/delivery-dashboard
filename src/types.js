// @flow
import type {
  Store as ReduxStore,
  ThunkAction as ReduxThunkAction,
  Dispatch as ReduxDispatch,
  GetState as ReduxGetState,
} from 'redux';

/*
 * state types
 */
export type OngoingVersion = [string, string];
export type OngoingVersions = OngoingVersion[];

export type Product = 'firefox';
export type Status = 'missing' | 'exists' | 'incomplete' | 'error';
export type Check =
  | 'archive'
  | 'product_details'
  | 'release_notes'
  | 'security_advisories'
  | 'download_links';

export type CheckInfo = {
  +url: string,
  +title: string,
};

export type ReleaseInfo = {
  +channel: string,
  +product: Product,
  +version: string,
  +checks: CheckInfo[],
};

export type CheckResult = {
  +status: Status,
  +message: string,
  +link: string,
};

export type CheckResults = {
  [check: string]: CheckResult,
};

export type APIVersionData = {
  name: string,
  version: string,
  source: string,
  commit: string,
};

export type Login = 'LOGGED_OUT' | 'LOGIN_REQUESTED' | 'LOGGED_IN';
export const LOGGED_OUT = 'LOGGED_OUT';
export const LOGIN_REQUESTED = 'LOGIN_REQUESTED';
export const LOGGED_IN = 'LOGGED_IN';

export type State = {
  +version: string,
  +versionInput: string,
  +latestChannelVersions: OngoingVersions,
  +releaseInfo: ?ReleaseInfo,
  +checkResults: CheckResults,
  +pollbotVersion: ?APIVersionData,
  +shouldRefresh: boolean,
  +login: Login,
  +userInfo: any,
};

/*
 * action types
 */
export const ADD_CHECK_RESULT = 'ADD_CHECK_RESULT';
export const SET_VERSION = 'SET_VERSION';
export const UPDATE_VERSION_INPUT = 'UPDATE_VERSION_INPUT';
export const SUBMIT_VERSION = 'SUBMIT_VERSION';
export const UPDATE_LATEST_CHANNEL_VERSIONS = 'UPDATE_LATEST_CHANNEL_VERSIONS';
export const UPDATE_RELEASE_INFO = 'UPDATE_RELEASE_INFO';
export const UPDATE_POLLBOT_VERSION = 'UPDATE_POLLBOT_VERSION';
export const UPDATE_USER_INFO = 'UPDATE_USER_INFO';

export type AddCheckResult = {|
  type: 'ADD_CHECK_RESULT',
  title: string,
  result: CheckResult,
|};
export type SetVersion = {|
  type: 'SET_VERSION',
  version: string,
|};
export type UpdateVersionInput = {|
  type: 'UPDATE_VERSION_INPUT',
  version: string,
|};
export type SubmitVersion = {|
  type: 'SUBMIT_VERSION',
|};
export type UpdateLatestChannelVersions = {|
  type: 'UPDATE_LATEST_CHANNEL_VERSIONS',
  versions: OngoingVersions,
|};
export type UpdateReleaseInfo = {|
  type: 'UPDATE_RELEASE_INFO',
  releaseInfo: ReleaseInfo,
|};
export type UpdatePollbotVersion = {|
  type: 'UPDATE_POLLBOT_VERSION',
  version: APIVersionData,
|};
export type LoggedIn = {|
  type: 'LOGGED_IN',
|};
export type LoggedOut = {|
  type: 'LOGGED_OUT',
|};
export type LoginRequested = {|
  type: 'LOGIN_REQUESTED',
|};
export type UpdateUserInfo = {|
  type: 'UPDATE_USER_INFO',
  userInfo: any,
|};

/*
 * saga types
 */
export const REQUEST_ONGOING_VERSIONS = 'REQUEST_ONGOING_VERSIONS';
export const REQUEST_POLLBOT_VERSION = 'REQUEST_POLLBOT_VERSION';
export const UPDATE_URL = 'UPDATE_URL';
export const REFRESH_STATUS = 'REFRESH_STATUS';
export const REQUEST_STATUS = 'REQUEST_STATUS';
export const REQUEST_LOGIN = 'REQUEST_LOGIN';
export const REQUEST_LOGOUT = 'REQUEST_LOGOUT';

export type RequestOngoingVersions = {|
  type: 'REQUEST_ONGOING_VERSIONS',
|};

export type RequestPollbotVersion = {|
  type: 'REQUEST_POLLBOT_VERSION',
|};

export type UpdateUrl = {|
  type: 'UPDATE_URL',
|};

export type RefreshStatus = {|
  type: 'REFRESH_STATUS',
|};

export type RequestStatus = {|
  type: 'REQUEST_STATUS',
  version: string,
|};

export type RequestLogin = {|
  type: 'REQUEST_LOGIN',
|};

export type RequestLogout = {|
  type: 'REQUEST_LOGOUT',
|};

export type Action =
  | AddCheckResult
  | LoggedIn
  | LoggedOut
  | LoginRequested
  | RefreshStatus
  | RequestLogin
  | RequestLogout
  | RequestOngoingVersions
  | RequestPollbotVersion
  | RequestStatus
  | SetVersion
  | SubmitVersion
  | UpdateLatestChannelVersions
  | UpdatePollbotVersion
  | UpdateReleaseInfo
  | UpdateUrl
  | UpdateUserInfo
  | UpdateVersionInput;

/*
 * Redux types
 */
export type GetState = ReduxGetState<State>;
export type ThunkAction<Result> = ReduxThunkAction<State, Action, Result>;
export type Store = ReduxStore<State, Action>;
export type Dispatch = ReduxDispatch<State, Action>;
