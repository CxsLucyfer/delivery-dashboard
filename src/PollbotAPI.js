// @flow
import type {
  APIVersionData,
  CheckResult,
  OngoingVersionsDict,
  ReleaseInfo,
} from './types';

export const SERVER = 'https://pollbot.dev.mozaws.net/v1';

export async function getOngoingVersions(): Promise<OngoingVersionsDict> {
  const response = await fetch(`${SERVER}/firefox/ongoing-versions`);
  return response.json();
}

export async function getReleaseInfo(version: string): Promise<ReleaseInfo> {
  const response = await fetch(`${SERVER}/firefox/${version}`);
  return response.json();
}

export async function checkStatus(url: string): Promise<CheckResult> {
  const response = await fetch(url);
  return response.json();
}

export async function getPollbotVersion(): Promise<APIVersionData> {
  const response = await fetch(`${SERVER}/__version__`);
  return response.json();
}
