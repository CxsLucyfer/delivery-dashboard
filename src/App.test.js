import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {mount, shallow} from 'enzyme';
import {
  App,
  ConnectedApp,
  Dashboard,
  DisplayStatus,
  LoginButton,
  parseUrl,
  SearchForm,
  versionInputDispatchProps,
} from './App';
import {Alert, Spin} from 'antd';
import createStore from './create-store';
import {SERVER} from './PollbotAPI';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {LOGGED_IN, LOGGED_OUT, LOGIN_REQUESTED} from './types';

Enzyme.configure({adapter: new Adapter()});

// Mock the Notification API.
global.Notification = {
  requestPermission: jest.fn(),
};

// Mock the localStorage API.
global.localStorage = (function() {
  var store = {};

  return {
    getItem: function(key) {
      return store[key] || null;
    },
    setItem: function(key, value) {
      store[key] = value.toString();
    },
    clear: function() {
      store = {};
    },
  };
})();

// Mock the Pollbot version (version won't be visible in the rendered
// component, as it's only visible after the state has been updated, not on
// first render.
function fetchMocker(response) {
  return jest
    .fn()
    .mockImplementation(() => Promise.resolve({json: () => response}));
}

global.fetch = fetchMocker({
  version: 'pollbot-version-number',
  commit: 'pollbot-commit-hash',
  source: 'https://github.com/mozilla/PollBot.git',
  name: 'pollbot',
});

// Mock the delivery-dashboard version
jest.mock('./version', () => ({
  version: 'version-number',
  commit: 'commit-hash',
  source: 'https://github.com/mozilla/delivery-dashboard.git',
  name: 'delivery-dashboard',
}));

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.clearAllTimers();
});

describe('<App />', () => {
  it('renders without crashing', () => {
    const app = renderer.create(
      <Provider store={createStore()}>
        <ConnectedApp />
      </Provider>,
    );
    expect(app.toJSON()).toMatchSnapshot();
  });
  it('requests for Notification permissions', () => {
    renderer.create(
      <Provider store={createStore()}>
        <ConnectedApp />
      </Provider>,
    );
    expect(global.Notification.requestPermission).toHaveBeenCalled();
  });
  it('requests PollBot for its version', () => {
    renderer.create(
      <Provider store={createStore()}>
        <ConnectedApp />
      </Provider>,
    );
    expect(global.fetch).toHaveBeenCalledWith(`${SERVER}/__version__`);
  });
  it('checks if the user just logged in', () => {
    const module = require('./auth0');
    module.checkLogin = jest.fn();

    const app = shallow(<App dispatch={jest.fn()} />).instance();
    expect(module.checkLogin).toHaveBeenCalledWith(app.onLoggedIn);
  });
  it('checks if the user is logged in', () => {
    const auth0Module = require('./auth0');
    auth0Module.isAuthenticated = jest.fn();
    const actionsModule = require('./actions');
    actionsModule.loggedIn = jest.fn();

    shallow(<App dispatch={jest.fn()} />);
    expect(auth0Module.isAuthenticated).toHaveBeenCalledTimes(1);
    expect(actionsModule.loggedIn).toHaveBeenCalledTimes(0);

    auth0Module.isAuthenticated = jest.fn(() => true);
    shallow(<App dispatch={jest.fn()} />);
    expect(auth0Module.isAuthenticated).toHaveBeenCalledTimes(1);
    expect(actionsModule.loggedIn).toHaveBeenCalledTimes(1);
  });
  it('dispatches loggedIn and requests user info', () => {
    const auth0Module = require('./auth0');
    auth0Module.fetchUserInfo = jest.fn();
    const actionsModule = require('./actions');
    actionsModule.loggedIn = jest.fn();

    const app = shallow(<App dispatch={jest.fn()} />).instance();
    const numCalls = actionsModule.loggedIn.mock.calls.length;
    app.onLoggedIn();
    expect(actionsModule.loggedIn).toHaveBeenCalledTimes(numCalls + 1);
    expect(auth0Module.fetchUserInfo).toHaveBeenCalledWith(app.onUserInfo);
  });
  it('dispatches userInfo', () => {
    const module = require('./actions');
    module.updateUserInfo = jest.fn();

    const app = shallow(<App dispatch={jest.fn()} />).instance();
    app.onUserInfo('foo');
    expect(module.updateUserInfo).toHaveBeenCalledWith('foo');
  });
  it('dispatches requestLogin', () => {
    const module = require('./actions');
    module.requestLogin = jest.fn();

    const app = shallow(<App dispatch={jest.fn()} />).instance();
    app.onLoginRequested();
    expect(module.requestLogin).toHaveBeenCalledTimes(1);
  });
  it('dispatches requestLogout', () => {
    const module = require('./actions');
    module.requestLogout = jest.fn();

    const app = shallow(<App dispatch={jest.fn()} />).instance();
    app.onLogoutRequested();
    expect(module.requestLogout).toHaveBeenCalledTimes(1);
  });
  it('calls requestStatus(version) with the version from the hash', () => {
    global.window.location.hash = '#pollbot/firefox/123.0';

    const module = require('./actions');
    module.requestStatus = jest.fn();

    const store = createStore();
    // We also need to mock the dispatch function, as it doesn't like to be
    // called with a mock.
    store.dispatch = jest.fn();

    renderer.create(
      <Provider store={store}>
        <ConnectedApp />
      </Provider>,
    );
    expect(module.requestStatus).toHaveBeenCalledWith('123.0');
  });
  it('sets up auto-refresh', () => {
    const module = require('./actions');
    module.requestStatus = jest.fn();
    module.refreshStatus = jest.fn();

    const wrapper = shallow(<App dispatch={jest.fn()} />);
    const app = wrapper.instance();
    app.stopAutoRefresh = jest.fn();

    // Called once, on mounting the component.
    jest.runOnlyPendingTimers();
    const numCalledRequestStatus = module.requestStatus.mock.calls.length;

    // Shouldn't auto-refresh => stop auto refresh.
    expect(app.stopAutoRefresh).toHaveBeenCalledTimes(0);
    wrapper.setProps({shouldRefresh: false});
    expect(app.stopAutoRefresh).toHaveBeenCalledTimes(1);
    expect(app.refreshIntervalId).toBeNull();
    app.setUpAutoRefresh();
    expect(app.stopAutoRefresh).toHaveBeenCalledTimes(2);
    expect(app.refreshIntervalId).toBeNull();
    jest.runOnlyPendingTimers();
    expect(module.requestStatus).toHaveBeenCalledTimes(numCalledRequestStatus);
    expect(module.refreshStatus).toHaveBeenCalledTimes(0);

    // Should auto-refresh => start auto refresh.
    expect(app.refreshIntervalId).toBeNull();
    wrapper.setProps({shouldRefresh: true});
    app.setUpAutoRefresh();
    expect(app.stopAutoRefresh).toHaveBeenCalledTimes(2); // Not called again.
    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(app.refreshIntervalId).toBeTruthy();
    jest.runOnlyPendingTimers();
    expect(module.requestStatus).toHaveBeenCalledTimes(numCalledRequestStatus);
    expect(module.refreshStatus).toHaveBeenCalledTimes(1);

    // Should auto-refresh, but already set up => don't start auto refresh.
    wrapper.setProps({shouldRefresh: true});
    app.setUpAutoRefresh();
    expect(app.stopAutoRefresh).toHaveBeenCalledTimes(2); // Not called again.
    expect(setInterval).toHaveBeenCalledTimes(1); // Not called again.
    expect(app.refreshIntervalId).toBeTruthy();
    expect(module.requestStatus).toHaveBeenCalledTimes(numCalledRequestStatus);
    expect(module.refreshStatus).toHaveBeenCalledTimes(1); // Not called again.
  });
  it('stops auto-refresh', () => {
    const app = shallow(<App dispatch={jest.fn()} />).instance();

    // Shouldn't call clearInterval if not needed.
    expect(app.refreshIntervalId).toBeNull();
    app.stopAutoRefresh();
    expect(clearInterval).toHaveBeenCalledTimes(0);
    // Should call clearInterval if needed.
    app.refreshIntervalId = 123;
    app.stopAutoRefresh();
    expect(clearInterval).toHaveBeenCalledWith(123);
    expect(app.refreshIntervalId).toBeNull();
  });
  it('stops the auto-refresh on unmount', () => {
    const wrapper = shallow(<App dispatch={jest.fn()} />);
    const app = wrapper.instance();
    app.stopAutoRefresh = jest.fn();
    wrapper.unmount();
    expect(app.stopAutoRefresh).toHaveBeenCalledTimes(1);
  });
});

describe('parseUrl', () => {
  it('returns null for a non matching url', () => {
    expect(parseUrl('')).toBeNull();
    expect(parseUrl('#foobar')).toBeNull();
  });
  it('returns the proper structure for a matching url', () => {
    expect(parseUrl('#pollbot/firefox/50.0')).toEqual({
      service: 'pollbot',
      product: 'firefox',
      version: '50.0',
    });
  });
});

describe('<SearchForm />', () => {
  it('handles input text change', () => {
    const module = require('./actions');
    module.updateVersionInput = jest.fn();

    const {handleSearchBoxChange} = versionInputDispatchProps(jest.fn());
    const wrapper = mount(
      <SearchForm handleSearchBoxChange={handleSearchBoxChange} />,
    );
    const input = wrapper.find('input');
    input.instance().value = 'foobar'; // Workaround for https://github.com/airbnb/enzyme/issues/218
    input.simulate('change', input);
    expect(module.updateVersionInput).toHaveBeenCalledWith('foobar');
  });
  it('handles dismissing a version', () => {
    const module = require('./actions');
    module.setVersion = jest.fn();

    const {handleDismissSearchBoxVersion} = versionInputDispatchProps(
      jest.fn(),
    );
    const wrapper = mount(
      <SearchForm
        handleDismissSearchBoxVersion={handleDismissSearchBoxVersion}
      />,
    );
    wrapper.find('.ant-input-group-addon i').simulate('click');
    expect(global.window.location.hash).toBe('');
    expect(module.setVersion).toHaveBeenCalledWith('');
  });
  it('handles the form submission', () => {
    const module = require('./actions');
    module.submitVersion = jest.fn();
    module.updateUrl = jest.fn();

    const {onSubmit} = versionInputDispatchProps(jest.fn());
    const wrapper = mount(<SearchForm onSubmit={onSubmit} />);
    wrapper.simulate('submit');
    expect(module.submitVersion).toHaveBeenCalled();
    expect(module.updateUrl).toHaveBeenCalled();
  });
});

describe('<Dashboard />', () => {
  const releaseInfo = {
    channel: 'nightly',
    product: 'firefox',
    version: '50.0',
    checks: [
      {url: 'some-url', title: 'some title'},
      {url: 'some-url-2', title: 'some title 2'},
    ],
  };
  const checkResults = {
    'some title': {
      status: 'exists',
      message: 'check is successful',
      link: 'some link',
    },
    'some title 2': {
      status: 'exists',
      message: 'check is successful',
      link: 'some link',
    },
  };
  it('displays a help text when no version is selected', () => {
    const wrapper = shallow(<Dashboard version="" />);
    expect(wrapper.text()).toContain('enter your version number');
  });
  it('displays a spinner when a version is selected', () => {
    const wrapper = shallow(<Dashboard version="50.0" />);
    expect(wrapper.find(Spin).length).toBe(1);
  });
  it('displays a list of check results when a release info is present', () => {
    const wrapper = shallow(
      <Dashboard
        version="50.0"
        releaseInfo={releaseInfo}
        checkResults={checkResults}
      />,
    );
    expect(wrapper.find(Spin).length).toBe(0);
    expect(wrapper.find(DisplayStatus).length).toBe(2);
  });
  it('displays a "complete" label when all the results are successful', () => {
    const wrapper = shallow(
      <Dashboard
        version="50.0"
        releaseInfo={releaseInfo}
        checkResults={checkResults}
        shouldRefresh={false}
      />,
    );
    const status = wrapper.find(Alert);
    expect(status.prop('message')).toEqual('complete');
    expect(status.prop('type')).toEqual('success');
  });
  it('displays an "incomplete" label if some results are unsuccessful', () => {
    const wrapper = shallow(
      <Dashboard
        version="50.0"
        releaseInfo={releaseInfo}
        checkResults={checkResults}
        shouldRefresh={true}
      />,
    );
    const status = wrapper.find(Alert);
    expect(status.prop('message')).toEqual('incomplete');
    expect(status.prop('type')).toEqual('error');
  });
});

describe('<DisplayStatus />', () => {
  const checkDisplayStatus = (status, label) => {
    const wrapper = mount(
      <DisplayStatus status={status} message="check message" url="check url" />,
    );
    const link = wrapper.find('a');
    expect(link.prop('href')).toEqual('check url');
    expect(link.prop('title')).toEqual('check message');
    const alert = wrapper.find(Alert);
    expect(alert.prop('type')).toBe(label);
    expect(link.text()).toEqual('check message');
  };
  it('displays the status when the status is exists', () => {
    checkDisplayStatus('exists', 'success');
  });
  it('displays the status when the status is incomplete', () => {
    checkDisplayStatus('incomplete', 'info');
  });
  it('displays the status when the status is missing', () => {
    checkDisplayStatus('missing', 'warning');
  });
  it('displays the error message when there an error', () => {
    checkDisplayStatus('error', 'error');
  });
});

describe('<Login Button />', () => {
  const onLoginRequested = jest.fn();
  const onLogoutRequested = jest.fn();
  it('displays a login icon and text when logged off', () => {
    const wrapper = mount(
      <LoginButton
        onLoginRequested={onLoginRequested}
        onLogoutRequested={onLogoutRequested}
        loginState={LOGGED_OUT}
      />,
    );
    const button = wrapper.find('button');
    expect(button.text()).toEqual('login');
    const icon = wrapper.find('i');
    expect(icon.prop('className')).toContain('login');

    expect(onLoginRequested).toHaveBeenCalledTimes(0);
    button.simulate('click');
    expect(onLoginRequested).toHaveBeenCalledTimes(1);
  });
  it('displays a logout icon and text when logged in', () => {
    const wrapper = mount(
      <LoginButton
        onLoginRequested={onLoginRequested}
        onLogoutRequested={onLogoutRequested}
        loginState={LOGGED_IN}
      />,
    );
    const button = wrapper.find('button');
    expect(button.text()).toEqual('logout');
    const icon = wrapper.find('i');
    expect(icon.prop('className')).toContain('logout');

    expect(onLogoutRequested).toHaveBeenCalledTimes(0);
    button.simulate('click');
    expect(onLogoutRequested).toHaveBeenCalledTimes(1);
  });
  it('displays a loading icon when login in', () => {
    const wrapper = mount(
      <LoginButton
        onLoginRequested={onLoginRequested}
        onLogoutRequested={onLogoutRequested}
        loginState={LOGIN_REQUESTED}
      />,
    );
    const button = wrapper.find('button');
    expect(button.text()).toEqual('login');
    const icon = wrapper.find('i');
    expect(icon.prop('className')).toContain('loading');
  });
});
