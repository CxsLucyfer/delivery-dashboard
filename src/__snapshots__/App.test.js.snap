// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders without crashing 1`] = `
<div>
  <header>
    <h1>
      <a
        href="."
      >
        Delivery Dashboard
      </a>
    </h1>
    <div
      className="user"
    >
      <button
        className="ant-btn"
        onClick={[Function]}
        type="button"
      >
        <i
          className="anticon anticon-login"
        />
        <span>
          login
        </span>
      </button>
    </div>
  </header>
  <div
    className="mainContent ant-layout ant-layout-has-sider"
  >
    <div
      className="ant-layout-sider"
      style={
        Object {
          "flex": "0 0 200px",
          "maxWidth": "200px",
          "minWidth": "200px",
          "width": "200px",
        }
      }
    >
      <div
        className="ant-layout-sider-children"
      >
        <div
          className="ant-spin ant-spin-spinning ant-spin-show-text"
        >
          <span
            className="ant-spin-dot"
          >
            <i />
            <i />
            <i />
            <i />
          </span>
        </div>
      </div>
    </div>
    <div
      className="ant-layout-content"
    >
      <form
        className="ant-form ant-form-horizontal"
        onSubmit={[Function]}
      >
        <div
          className="ant-row ant-form-item"
          style={undefined}
        >
          <div
            className="ant-form-item-control-wrapper"
          >
            <div
              className="ant-form-item-control "
            >
              <span
                className="ant-input-group-wrapper"
                style={undefined}
              >
                <span
                  className="ant-input-wrapper ant-input-group"
                >
                  <input
                    className="ant-input"
                    disabled={false}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Firefox version, eg. \\"57.0\\""
                    style={null}
                    type="search"
                    value=""
                  />
                  <span
                    className="ant-input-group-addon"
                  >
                    <i
                      className="anticon anticon-close"
                      onClick={[Function]}
                      style={
                        Object {
                          "cursor": "pointer",
                        }
                      }
                    />
                  </span>
                </span>
              </span>
            </div>
          </div>
        </div>
      </form>
      <p>
        Learn more about a specific version.
        <strong>
           Select or enter your version number.
        </strong>
      </p>
    </div>
  </div>
  <footer>
    Delivery dashboard version:
     
    <a
      href="https://github.com/mozilla/delivery-dashboard/commit/commit-hash"
    >
      version-number
    </a>
     -- Pollbot version:
     
  </footer>
</div>
`;
